### PR TITLE
swarm/network: remove dead code

### PR DESCRIPTION
**Pull request by nonsense**
This PR is:
1. Fixing message handlers to run asynchronously so that we don't block the TCP socket for a given peer.
2. Fixing the signature for peer.Drop, which should not take an error as it doesn't do anything with that argument.
3. Remove dead code - `SwarmChunkServer`
4. Remove dead code - partly removing `skipCheck` - this will be a longer effort
5. Remove the `RETRIEVE_REQUEST` stream, which is not used in any production code for anything.
6. Remove tests that somehow expect an OfferedHashes messages after a RetrieveRequest? This is definitely not how `master` code has been working for months.

### CODE REVIEW

File: swarm/network/stream/delivery.go
Comment (nonsense) [inline swarm/network/stream/delivery.go:156]: One of the tests is actually sending a `ChunkDeliveryMsg` and not the other two specific types, therefore failing.

This is not code that is hit in production, but I decided to add this here temporarily, rather than dig into the test.

It is not clear why this test was not failing with the previous code.

File: swarm/network/stream/delivery_test.go
Comment (nonsense) [inline swarm/network/stream/delivery_test.go:66]: `SwarmChunkServer` doesn't exist anymore, and `RETRIEVE_REQUEST` doesn't exist anymore.

Also this is a huge test, which only checks that if we issue a request to `RequestFromPeers`, we get specific messages back - `SubscribeMsg` for the `RETRIEVE_REQUEST` stream, and a `RetrieveRequestMsg` - this is rather useless for two reasons:
1. We are locking behaviour that doesn't provide much value.
2. We are testing a very basic functionality, which is hit 1000 times per a single integration test.

No need to lock this behaviour in such a complicated way. Not to mention that `RETRIEVE_REQUEST` stream serves no purpose in any production code for a long time.

File: swarm/network/stream/delivery_test.go
Comment (nonsense) [inline swarm/network/stream/delivery_test.go:268]: How did that work before is beyond me... We don't have any validation for `ChunkDeliveryMsg` ?

File: swarm/network/stream/delivery.go
Comment (janos) [inline swarm/network/stream/delivery.go:168]: streamer.quit is canceling context in this goroutine on shutdown. If that is removed, then this goroutine is not doing anything as calling cancel on context that is done has no effect.

File: swarm/network/stream/delivery.go
Comment (janos) [inline swarm/network/stream/delivery.go]: Should this osp log field be removed also?

File: swarm/network/stream/delivery.go
Comment (janos) [inline swarm/network/stream/delivery.go]: It is be better to use named priority as the third argument.

File: swarm/network/stream/stream.go
Comment (janos) [inline swarm/network/stream/stream.go:96]: RetrievalOption type, its constants, Registry.autoRetrieval, RegistryOptions.Retrieval and references in tests can be removed, since this values are not used anymore.

File: swarm/network/stream/stream.go
Comment (janos) [inline swarm/network/stream/stream.go]: This function should return on <-r.quit to terminate the goroutine in which it runs.

File: swarm/network/stream/stream.go
Comment (janos) [inline swarm/network/stream/stream.go:581]: The error should be logged.

Since most of the messages are creating a goroutine, would it be simpler to have a goroutine with the whole switch statement in it and remove the repetitive code?

File: swarm/network/stream/stream.go
Comment (janos) [inline swarm/network/stream/stream.go:601]: The error should be logged.

File: swarm/network/stream/stream.go
Comment (janos) [inline swarm/network/stream/stream.go:612]: The error should be logged.

File: swarm/network/stream/stream.go
Comment (janos) [inline swarm/network/stream/stream.go:623]: The error should be logged.

File: swarm/network/stream/stream.go
Comment (janos) [inline swarm/network/stream/stream.go:633]: The error should be logged.

File: swarm/network/stream/streamer_test.go
Comment (janos) [inline swarm/network/stream/streamer_test.go]: TestGetSubscriptions-> TestGetServerSubscriptions

File: swarm/network/stream/streamer_test.go
Comment (janos) [inline swarm/network/stream/streamer_test.go:1323]: stream_getPeerSubscriptions -> stream_getPeerServerSubscriptions

File: swarm/network/stream/stream.go
Comment (janos) [inline swarm/network/stream/stream.go]: Maybe to write a test for this function as well?

**Review by janos — COMMENTED** [review]
There is a lot of work in this PR, thanks Anton.

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go]: This function is trivial, just like the `GetPeerServerSubscriptions`. I don't think we need tests for it, chances that someone will refactor it and break it are so slim, that I'd rather we don't add 50 lines of tests like the other one.

Also this is not an API call that is part of critical functionality, it is here for debug purposes.

If we add unit test for this, we will probably have to remove it, once we review the `Stream` implementation - something that has been in discussions over the last weeks.

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:623]: done

File: swarm/network/stream/delivery.go
Comment (nonsense) [reply swarm/network/stream/delivery.go]: done

File: swarm/network/stream/delivery.go
Comment (nonsense) [reply swarm/network/stream/delivery.go]: You're right. Priority queue will be removed in a follow up PR, but since it is not here, I am just using `Top`. Let me know if this priority makes sense.

To me the priority queue is redundant and we don't have any proof that it is actually needed and not premature optimization. I've asked multiple times about this, and have never received a straight answer about it.

File: swarm/network/stream/delivery.go
Comment (nonsense) [reply swarm/network/stream/delivery.go:168]: Any suggestions how to fix this? Because of `RETRIEVE_REQUEST` stream, we no longer have `s` and we no longer have `streamer.

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:96]: Is it okay to do that in a follow up PR, or you want it done here?

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go]: done

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:581]: Error is logged now, you're probably looks at an outdated diff.

The repetitive code is just a log line and a p.Drop(), I don't have preference if it should be in an external function or not.

File: swarm/network/stream/streamer_test.go
Comment (nonsense) [reply swarm/network/stream/streamer_test.go]: done

File: swarm/network/stream/streamer_test.go
Comment (nonsense) [reply swarm/network/stream/streamer_test.go:1323]: done

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go]: Actually it is not 50 lines, but much more, we also have `TestGetServerSubscriptionsRPC`

File: node/config.go
Comment (nonsense) [reply node/config.go]: reverted

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:633]: done

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go]: `TestGetServerSubscriptionsRPC` depends on:
- `simulations` package
- `adapters`
- `netStore`
- `delivery`
- spec for `SubscribeMsg`
- json snapshots
- exact naming of the function, not covered by the compiler by dynamically linked

I'd rather we don't write such low-value tests for such trivial functionalities.

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go:168]: I think that Delivery should have a Close method to terminate its goroutines.

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go]: I think that Mid priority is 1 and Top is 3. If you plan to remove the priority queue, then it does not matter so much.

File: swarm/network/stream/stream.go
Comment (janos) [reply swarm/network/stream/stream.go:96]: Since they are all related to the retrieval which is removed in this PR, I think that they should be removed here, too.

File: swarm/network/stream/stream.go
Comment (janos) [reply swarm/network/stream/stream.go:581]: Thanks for the log line. Maybe others can express their opinions about repetition in this function, it is working, but I do not find it nice.

File: swarm/network/stream/delivery.go
Comment (nonsense) [reply swarm/network/stream/delivery.go:168]: @janos I agree with you, but I think this is a feature we can add later. If I understand correctly, you want to have a clean node process which you will be able to start/stop in simulations and which handles the lifecycle of its go-routine correctly.

In a long-running Swarm process right now, I'd say that it doesn't matter if the go-routines are terminated when Delivery exits, because the main thread will exit.

I'd rather we focus on the rest of the fixes and bugs that we have already uncovered and solve this later - namely get the rest of the changes from `simplify-fetchers` in, and then address this.

We can focus on getting the gist of the simplified fetchers refactor in and address the subscription bug, so that we can release v0.4 or we can focus on every detail that is not handled correctly, and get there slower. I think we should do the former, but this is up for discussion :)

File: swarm/network/stream/stream.go
Comment (janos) [reply swarm/network/stream/stream.go]: Well, tools that can be used for testing the code should not be relevant for decision if the code should be tested. If this code exists and somebody refactors the stream package, I suppose that it would be nice to have a test to verify that there is no regression. I would not block this PR for not testing this code if it is not an issue for others.

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go:168]: I think that this is not a large feature, and that such things we should not leave unsolved. It is very important to correctly handle resources. If we skip on this types of resource handling, we will create more technical dept.

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:96]: removed.

Comment (janos) [reply swarm/network/stream/stream.go:96]: Thanks Anton

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go]: @janos I strongly disagree with this.

If you are using 10 libraries to test a `for` loop that is adding values to a slice, you are increasing the complexity of the codebase for no good reason.

Now that these tests are part of the codebase, every time we change any of those 10 libraries, we also have to modify the tests. This is a cost to development that we should not ignore. The benefit of this test is that we are sure that we have not broken the `for` loop that is adding values to a slice - rather low-value in my opinion.

File: swarm/network/stream/delivery.go
Comment (nonsense) [reply swarm/network/stream/delivery.go:168]: done

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:581]: The mental load that this puts to the reader is close to 0. It really doesn't matter. We should focus on other parts of Swarm, but this.

Comment (janos) [reply swarm/network/stream/stream.go:581]: You are right about the mental load, but my opinion is that this function can be structured to be easier to maintain. I think that every part of the swarm is important equally if we want to have a better code. But it is not a big deal, as I said, it works.

File: swarm/network/stream/stream.go
Comment (janos) [reply swarm/network/stream/stream.go]: Thanks for explanations. I will not insist on this test.

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go:168]: Thanks Anton.

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go]: @janos giving a bit more thought to this particular piece of code - the sole purpose of this API call exists is to be a helper method to debug the already broken subscriptions implementation.

I'd rather we focus on solving the bugs in the subscriptions implementation (without commenting what this involves - rewrite or bug fix or whatever), than write tests for the helper methods we add to the codebase to help us track the bigger issues in Swarm.

This feels like adding a test for a metrics counter or timer to me.

So if we insist to have a test on this API (which I see you don't in the comment above), I'd rather just remove it altogether, and find another way to debug the subscriptions functionality.

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:581]: Handling incoming messages asynchronously should probably be part of the `p2p/protocols` abstraction, which is not really in scope for this PR.

In light of our `guidelines` and `standards`, I think that this PR is increasing the code quality from the previous state of the code, because:
- a lot of dead code is being removed
- a bug is being solved

Therefore I suggest we address the repetition of the `go log error drop` flow, since we have a few other bugs that are not yet integrated to the codebase.

**Review by janos — APPROVED** [review]
Thanks for addressing my comments Anton.

**Comment by nonsense** [comment]
@janos thanks for thorough review and for putting up with my rushed PR!

File: swarm/network/stream/delivery_test.go
Comment (zelig) [reply swarm/network/stream/delivery_test.go:268]: further up in localstore i guess there is. not on msg level

File: p2p/protocols/protocol.go
Comment (zelig) [inline p2p/protocols/protocol.go:246]: can we please preserve the error and log it or even put it in p2p.DiscSubprotocolError?
this is losing too much contextual info

**Review by zelig — COMMENTED** [review]
are you sure light node functionality does not regress here?

changes here are maybe too big, too hasty ?

**Review by zelig — COMMENTED** [review]
is removing the retrieve request stream justified because in current production code skipcheck is true. There was a reason why skipcheck was a (tested) option for retrieve request stream: we wanted to check if in some scenarios, chunk delivery for retrieve requests  should go via the offered/wanted hash roundtrip. While this may introduce latency, without it multiple successful requests will lead to multiple deliveries. 

At this point I am leaning towards handling the latter issue with request cancellation https://github.com/ethersphere/go-ethereum/issues/409 which also has the chance to cancel requests for all forwarding nodes 

In the light of these changes, the question arises: why are retrieve requests even part of the stream package and stream protocol. it would make sense to remove request handling entirely from stream, put it under network and make it part of `bzz` protocol directly.

File: swarm/network/stream/delivery_test.go
Comment (nonsense) [reply swarm/network/stream/delivery_test.go:268]: true, there is validation in localstore, i just found this confusing here. now that you mention it, it makes sense.

File: p2p/protocols/protocol.go
Comment (nonsense) [reply p2p/protocols/protocol.go:246]: We could log the error, but I'd rather not propagate it down to p2p, and replace `DiscSubprotocolError` just now, as this is changing behaviour, and we already have too many changes.

The current change preserves the behavior we already have.

**Comment by nonsense** [comment]
@zelig good question. I think code that is not used by the users, but only hit in unit tests, has no place in the codebase as a general rule of thumb.

Currently we have:
- RETRIEVE_REQUEST stream (retrieve requests handled through an offered/wanted hashes flow?),
- `SwarmChunkServer`
- skipCheck
- priority queue
- TakeoverProof

None of this is necessary for core Swarm functionality and is increasing the `carry` cost of development, and making code very difficult to reason about. This is the primary reason we have had bugs in the most simple use-cases of Swarm - namely upload a file and try to fetch it from a different node in the network.

To summarize, the bugs we've found during this effort, that were 'hidden' due to this non-used functionality:
- blocking TCP sockets, hidden behind `peer.Drop()` and hidden behind priority queue sends.
- an order of magnitude more ChunkDelivery messages due to the offered-hashes/wanted-hashes protocol not keeping track of them
- (not a known bug, but very convoluted implementation of the default RetrieveRequest flow)
- impossibility to add instrumentation and see Swarm behavior in a real network due to priority queues and convoluted use of contexts.

Because of all this I think it is justified to remove this dead code.

---

Regarding the light client - it is possible that this is breaking a very small part of it, but fixing it should be rather simple.

**Comment by nonsense** [comment]
RetrieveRequests in the simplified version of the code, have nothing to do with Streams. The only thing they share is the Fetchers, so that we have only 1 request per chunk, no matter if this request was done through a RetrieveRequest or through Syncing.

You have a point that this doesn't even belong in this package, but I have a hard time getting even simpler changes in, so I'd rather we change this gradually.

**Comment by janos** [comment]
Moving retrieve requests from the stream package is a very good idea. It would just need a different protocol as it is now part of the stream protocol specification.

**Comment by nonsense** [comment]
Yes, @janos, that's correct. Let's postpone this until we have all the known bug fixes integrated in the codebase, together with a solution for the subscription bug. Refactoring RetrieveRequests out of the Stream protocol is rather low priority right now, in my opinion.

**Comment by janos** [comment]
I agree, @nonsense.

File: swarm/network/stream/lightnode_test.go
Comment (acud) [inline swarm/network/stream/lightnode_test.go:26]: I think that the `registryOptions` is supposed to indicate the this is a light node behavior but I see what you mean, it is not so clear what is being tested here

File: swarm/network/stream/lightnode_test.go
Comment (acud) [inline swarm/network/stream/lightnode_test.go:65]: not sure what this is testing without testing either

File: swarm/network/stream/stream.go
Comment (acud) [inline swarm/network/stream/stream.go:105]: isn't this somehow necessary for light nodes too?

File: swarm/network/stream/stream.go
Comment (acud) [reply swarm/network/stream/stream.go:581]: i am good with leaving these functions as is right now for clarity.
if we keep the error (and actually use it) in `peer.Drop` we could drop the log because the error will be automatically logged on the peer drop

**Review by acud — COMMENTED** [review]
@nonsense great work. i left a few minor comments

File: p2p/protocols/protocol.go
Comment (nonsense) [reply p2p/protocols/protocol.go:246]: @justelad I am not changing this behavior as well in this PR, let's keep it as small as possible.

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:105]: Good question. You see which tests are being removed, and all other tests are passing.

**Comment by nonsense** [comment]
There are many discussions about `peer.Drop()` in this PR. I want to highlight that we are failing with developing Swarm in the most basic use-case when all the network is of `trusted` nodes, that adhere to the protocols.

Adding `peer.Drop()` at various locations and reconnecting right after that to the same peer is already broken and wrong. If the peer is misbehaving, why are we re-connecting to them right after we drop them?

Additionally `peer.Drop()` is a very insidious way to cover for deadlocks and logical problems with the protocols - if a peer is `misbehaving` (read `deadlocked`) and it times-out, we just disconnect and reconnect and move on, essentially solving the deadlock at an expense of a timeout.

We should first make sure that Swarm works well in a controlled environment, and then add tests that cover `misbehaving` peers. We currently see what happens if you try to do both of these things simultaneously.

File: swarm/network/stream/lightnode_test.go
Comment (nonsense) [reply swarm/network/stream/lightnode_test.go:65]: it seems to test that if `RETRIEVE_REQUEST` stream is not registered, then we should get an error. Not sure how that is valuable at all, since I've never see any actual usage of `RETRIEVE_REQUEST` stream.

I also don't see any light node tests other than these, so it is kinda hard to say if we are breaking anything or not.

File: swarm/network/stream/stream.go
Comment (nonsense) [reply swarm/network/stream/stream.go:105]: It would have been nice to have at least one test that starts a `lightnode` client, a `regular` client and fetches content from the lightnode. I am not aware of such a test, and since Travis is passing, I assume that we are not breaking anything.

**Comment by acud** [comment]
OK I see where you're going with this and I tend to agree. I never understood why we need to drop a peer (the p2p package negotiation should already cover for protocol version discrepancies) without actually maintaining a peer blacklist (whether if a decaying one to mitigate DoS vectors or a permanent one for litigation/other cryptoeconomic violations)

**Review by acud — APPROVED** [review]

**Comment by nonsense** [comment]
@justelad yes, dropping peers, without maintaining a blacklist also doesn't make sense to me. Maybe your node is misbehaving because it has a temporary network problem, that could be one reason where you might want to ignore a node for a while, but try them again later.

Bottom line - this functionality needs more work (backoff, blacklists, etc.), and in its current form it is more harmful than useful.

**Comment by zelig** [comment]
@nonsense i must say i agree with everything you responded.
great job...
I will keep insisting with certain bits that falls under 'feature regression', but it will be much nicer to add them and check their impact or prove them pointless :)

**Comment by nonsense** [comment]
PR has two approvals, so merging for now. This is not going to `master` in the foreseeable future, so we have plenty of time to decide if some of the `dead code` is actually not dead code, but useful (I doubt that). One thing to review is the light client, which AFAIK has 0 integration tests... probably worth to run the binary with `--lightnode` flag just to see if it works or not.

However I think it is more important to integrate the rest of the bugfixes, so that we have a more stable full swarm node, and only then worry about the rest.

**Comment by nonsense** [comment]
@zelig about the pointlessness of the features - the latencies that we have on the `simplify-fetchers` branch are much lower than those on `master`, so this of itself is a proof we are going in the right direction.

I very much welcome any improvements on those latencies - whether through a benchmark test or through an experiment between two deployments, or some other way, but definitely not just based on intuition, unless it is very easy argument to win everyone on the team on.

**Pull request merged**
